### ISSUE TITLE
On boot, some, not all, network configuration units are "Stopped" and then immediately "Started" again for no reason

### ISSUE THREAD

**Issue by thx1111** [systemd/systemd#31884]
### systemd version the issue has been seen with

255.4-2

### Used distribution

Arch

### Linux kernel version used

6.8.1.arch1-1

### CPU architectures issue was seen on

i686

### Component

systemd

### Expected behaviour you didn't see

Units start only once and are not stopped and started again for no reason.

### Unexpected behaviour you saw

On boot, some, not all, network configuration units are "Stopped" and then immediately "Started" again for no reason.  One unit runs the WIDE dhcp6c as Type= simple.  Another unit sets-up a sit tunnel as Type= oneshot.  TimeoutStartSec= 180s has been set explicitly, just as a check.  No other units see this problem.  In particular, the similar IPv4 udhcpc unit, running on the same WAN interface, does *not* see this problem.
```
Mar 20 22:17:00 topaz systemd[1]: Created slice Slice /system/dhcp6c-enp4s0.
Mar 20 22:17:00 topaz systemd[1]: Reached target ll.target.
Mar 20 22:17:07 topaz systemd[1]: Starting WIDE-DHCPv6 dhcp6c with PD Subnet on enp4s0 and NA on interface enp6s0...
Mar 20 22:17:14 topaz systemd[1]: Started WIDE-DHCPv6 dhcp6c with PD Subnet on enp4s0 and NA on interface enp6s0.
Mar 20 22:17:51 topaz systemd[1]: Stopping WIDE-DHCPv6 dhcp6c with PD Subnet on enp4s0 and NA on interface enp6s0...
Mar 20 22:18:01 topaz systemd[1]: Stopped WIDE-DHCPv6 dhcp6c with PD Subnet on enp4s0 and NA on interface enp6s0.
Mar 20 22:18:01 topaz systemd[1]: Starting WIDE-DHCPv6 dhcp6c with PD Subnet on enp4s0 and NA on interface enp6s0...
Mar 20 22:18:10 topaz systemd[1]: Started WIDE-DHCPv6 dhcp6c with PD Subnet on enp4s0 and NA on interface enp6s0.
```
After the second "Start", the unit is left alone to run.
```
$ systemd-analyze dot 'dhcp*'
digraph systemd {
        "shutdown.target"->"dhcp6c-enp4s0@enp6s0.service" [color="green"];
        "network.target"->"dhcp6c-enp4s0@enp6s0.service" [color="green"];
        "sys-subsystem-net-devices-enp6s0.device"->"dhcp6c-enp4s0@enp6s0.service" [color="grey66"];
        "dhcp6c-enp4s0@enp6s0.service"->"systemd-journald.socket" [color="green"];
        "dhcp6c-enp4s0@enp6s0.service"->"ll.target" [color="green"];
        "dhcp6c-enp4s0@enp6s0.service"->"sys-subsystem-net-devices-enp6s0.device" [color="green"];
        "dhcp6c-enp4s0@enp6s0.service"->"sysinit.target" [color="green"];
        "dhcp6c-enp4s0@enp6s0.service"->"system-dhcp6c\x2denp4s0.slice" [color="green"];
        "dhcp6c-enp4s0@enp6s0.service"->"basic.target" [color="green"];
        "dhcp6c-enp4s0@enp6s0.service"->"system-dhcp6c\x2denp4s0.slice" [color="black"];
        "dhcp6c-enp4s0@enp6s0.service"->"sysinit.target" [color="black"];
        "dhcp6c-enp4s0@enp6s0.service"->"ll.target" [color="black"];
        "dhcp6c-enp4s0@enp6s0.service"->"shutdown.target" [color="red"];
}
   Color legend: black     = Requires
                 dark blue = Requisite
                 dark grey = Wants
                 red       = Conflicts
                 green     = After
```
The similar udhcp unit is not effected by this mysterious "Stop" on boot:
```
$ systemd-analyze dot 'udhcp*'
digraph systemd {
        "shutdown.target"->"udhcpc@enp6s0.service" [color="green"];
        "sit1@enp6s0.service"->"udhcpc@enp6s0.service" [color="green"];
        "network.target"->"udhcpc@enp6s0.service" [color="green"];
        "sys-subsystem-net-devices-enp6s0.device"->"udhcpc@enp6s0.service" [color="grey66"];
        "udhcpc@enp6s0.service"->"systemd-journald.socket" [color="green"];
        "udhcpc@enp6s0.service"->"ll.target" [color="green"];
        "udhcpc@enp6s0.service"->"system-udhcpc.slice" [color="green"];
        "udhcpc@enp6s0.service"->"basic.target" [color="green"];
        "udhcpc@enp6s0.service"->"sysinit.target" [color="green"];
        "udhcpc@enp6s0.service"->"sysinit.target" [color="black"];
        "udhcpc@enp6s0.service"->"system-udhcpc.slice" [color="black"];
        "udhcpc@enp6s0.service"->"ll.target" [color="black"];
        "udhcpc@enp6s0.service"->"shutdown.target" [color="red"];
}
   Color legend: black     = Requires
                 dark blue = Requisite
                 dark grey = Wants
                 red       = Conflicts
                 green     = After
```
And, for the sit tunnel:
```
Mar 20 22:17:00 topaz systemd[1]: Created slice Slice /system/sit1.
Mar 20 22:17:08 topaz systemd[1]: Starting Tunnel Interface "sit1" through Physical Interface enp6s0...
[  There is no "Stopping ..." statement in the journal log for this unit. ]
Mar 20 22:17:16 topaz systemd[1]: sit1@enp6s0.service: Control process exited, code=killed, status=15/TERM
Mar 20 22:17:16 topaz systemd[1]: sit1@enp6s0.service: Failed with result 'signal'.
Mar 20 22:17:16 topaz systemd[1]: Stopped Tunnel Interface "sit1" through Physical Interface enp6s0.
Mar 20 22:17:16 topaz systemd[1]: Starting Tunnel Interface "sit1" through Physical Interface enp6s0...
Mar 20 22:17:51 topaz systemd[1]: Finished Tunnel Interface "sit1" through Physical Interface enp6s0.
```
Again, this unit is left alone after the second "Start".
```
$ systemd-analyze dot 'sit1*'
digraph systemd {
        "shutdown.target"->"sit1@enp6s0.service" [color="green"];
        "sit1@enp6s0.service"->"sysinit.target" [color="green"];
        "sit1@enp6s0.service"->"system-sit1.slice" [color="green"];
        "sit1@enp6s0.service"->"sys-subsystem-net-devices-enp6s0.device" [color="green"];
        "sit1@enp6s0.service"->"basic.target" [color="green"];
        "sit1@enp6s0.service"->"syslog.target" [color="green"];
        "sit1@enp6s0.service"->"systemd-journald.socket" [color="green"];
        "sit1@enp6s0.service"->"udhcpc@enp6s0.service" [color="green"];
        "sit1@enp6s0.service"->"sysinit.target" [color="black"];
        "sit1@enp6s0.service"->"system-sit1.slice" [color="black"];
        "sit1@enp6s0.service"->"shutdown.target" [color="red"];
        "sys-subsystem-net-devices-enp6s0.device"->"sit1@enp6s0.service" [color="grey66"];
}
   Color legend: black     = Requires
                 dark blue = Requisite
                 dark grey = Wants
                 red       = Conflicts
                 green     = After
```

### Steps to reproduce the problem

system boot

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
See above.
```

**Comment by yuwata** [comment]
Please show relevant .service units.

**Comment by poettering** [comment]
Closing due to lack of response

**Comment by thx1111** [comment]
Ok, I realized that, though  systemd "debug" on boot still causes a boot failure, the same mystery "kill" and "stop" occurs when the ethernet interface driver module is removed and reloaded.  So now, boot normally, then `kill -56 1`, `modprobe -r sky2;modprobe sky2`, and then let's look through the log, and take these one at a time, beginning where sit1@enp6s0.service is mysteriously killed.  Maybe just some excerpts for now, and tell me if you need more.

sit1@enp6s0.service has
```
After= sys-subsystem-net-devices-%i.device
BindsTo= sys-subsystem-net-devices-%i.device
ConditionPathExists= /sys/class/net/%i

[Service]
EnvironmentFile= /etc/conf.d/network.conf
Type= oneshot
RemainAfterExit= yes
...
```
and then runs a sequence of shell commands to set-up sit1.

Then in the log, there is:
```
 systemd[1]: sit1@enp6s0.service: Changed dead -> start-pre
 systemd[1]: Starting Tunnel Interface "sit1" through Physical Interface enp6s0...
...
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1147 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1148 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
...
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1159 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1160 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
...
 systemd[1]: sit1@enp6s0.service: Trying to enqueue job sit1@enp6s0.service/restart/replace
 systemd[1]: sit1@enp6s0.service: Merged into running job, re-running: sit1@enp6s0.service/restart as 7268
 systemd[1]: sit1@enp6s0.service: Enqueued job sit1@enp6s0.service/restart as 7268
 systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/job/7268 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/job/7268 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1219 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=method_return sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=2 reply_cookie=1 signature=o error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=3 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=4 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1220 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1221 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: sit1@enp6s0.service: Changed start-pre -> stop-sigterm
 systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=5 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=6 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1222 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sit1_40enp6s0_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1223 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1/job/7268 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/job/7268 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=1224 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
 systemd[1]: Received SIGCHLD from PID 1866 (sh).
 systemd[1]: Child 1866 (sh) died (code=killed, status=15/TERM)
 systemd[1]: sit1@enp6s0.service: Child 1866 belongs to sit1@enp6s0.service.
 systemd[1]: sit1@enp6s0.service: Control process exited, code=killed, status=15/TERM
 systemd[1]: sit1@enp6s0.service: Got final SIGCHLD for state stop-sigterm.
 systemd[1]: sit1@enp6s0.service: Will spawn child (service_enter_stop_post): /usr/bin/sh
```
Obvious questions -

Why is systemd "Trying to enqueue job sit1@enp6s0.service/restart/replace" when the job is already running?  And then, why has systemd "Enqueued job sit1@enp6s0.service/restart as 7268", for an already running job?

And the consequence apparently is "sit1@enp6s0.service: Changed start-pre -> stop-sigterm", for what appears to be no reason. `ExecStartPre=` is a wait loop, waiting for a good network connection to the remote host from a successful `ping`.

**Comment by thx1111** [comment]
See also https://github.com/systemd/systemd/issues/33969

**Comment by yuwata** [comment]
Not sure, but BindsTo= causes the restart, I guess.
Please try `udevadm trigger /sys/class/net/enp6s0` and check if the service is restarted.

**Comment by yuwata** [comment]
Note, typically `Wants=` and `After=` for the device unit is enough, and `BindsTo=` is not necessary. Though, I have no idea why you use `BindsTo=`. If `BindsTo=` causes your issue, then try to consider replacing `BindsTo=` with `Wants=`.

**Comment by thx1111** [comment]
Thanks for your note.  Hmm - `man systemd.unit` tells me:

```
BindsTo=
           Configures requirement dependencies, very similar in style to Requires=. However, this dependency type is stronger: in addition to the effect of Requires= it declares that if the unit bound to is stopped, this unit
           will be stopped too. This means a unit bound to another unit that suddenly enters inactive state will be stopped too. Units can suddenly, unexpectedly enter inactive state for different reasons: the main process of a
           service unit might terminate on its own choice, the backing device of a device unit might be unplugged or the mount point of a mount unit might be unmounted without involvement of the system and service manager.

           When used in conjunction with After= on the same unit the behaviour of BindsTo= is even stronger. In this case, the unit bound to strictly has to be in active state for this unit to also be in active state. This not
           only means a unit bound to another unit that suddenly enters inactive state, but also one that is bound to another unit that gets skipped due to an unmet condition check (such as ConditionPathExists=,
           ConditionPathIsSymbolicLink=, ... — see below) will be stopped, should it be running. Hence, in many cases it is best to combine BindsTo= with After=.
```
As I understand, this says that, *without* `BindsTo=` device interface, when the device interface disappears, for whatever reason, that the associated service units would *not* be "Stopped", and then also, that the included `ExecStopPost=` routines would not be run.  And that is not acceptable.

If you know differently, please explain.

Also, to be clear, note that:
```
systemd[1]: Stopped Tunnel Interface "sit1" through Physical Interface enp6s0.
```
*preceeds* `/usr/bin/modprobe sky2`.  There is no other `sit1@enp6s0.service` still running when the network interface appears again.

**Comment by yuwata** [comment]
Does `udevadm trigger` actually cause the restart of the unit?

**Comment by yuwata** [comment]
BTW, could you also provide relevant debugging logs for the device unit?

**Comment by yuwata** [comment]
With v256.11, I cannot reproduce the issue with the following unit and `udevadm trigger`. I also tried `systemctl daemon-reload` or `systemctl daemon-reexec` but still cannot reproduced.
```ini
[Unit]
After=sys-subsystem-net-devices-hoge.device
BindsTo=sys-subsystem-net-devices-hoge.device

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStartPre=sleep 5m
ExecStart=echo start
ExecStop=echo stop
ExecStopPost=echo stop-post
```

**Comment by yuwata** [comment]
Please try to create minimal reproducer of the issue.

**Comment by thx1111** [comment]
> Not sure, but BindsTo= causes the restart, I guess.

"I guess" is really why this is two separate issues.  One issue is understanding why `systemd` has "enqueued" a unit that is already running.  The second issue, though, is that `systemd` fails to report *why* it does what it does, even in "debug" mode.

But then, maybe `systemd` didn't report "why" it enqued a unit that is already running because it wasn't suppose to do that in the first place, and there's a bug in the state engine.

I'll poke around some more.

**Comment by thx1111** [comment]
> Does udevadm trigger actually cause the restart of the unit?

Apparently, no.  I've run `sudo udevadm trigger -v -t change -w sys-subsystem-net-devices-enp6s0.device`, `sudo udevadm trigger -v -c add -w sys-subsystem-net-devices-enp6s0.device`, and simply `sudo udevadm trigger /sys/class/net/enp6s0`, as suggested.  There is subsequently no response in the logs from any of the service units bound to the network interface enp6s0, including from `sit1@enp6s0.service`.

I don't understand the intent there, since `udevadm trigger` does not accurately represent the state transitions of interest.  Consider: 1) the network interface unit and service units are already active when the trigger occurs, but there is nothing to do; 2) the network interface unit and service units are inactive, but all the service units would fail if activated by the trigger simply because there is no actual network interface to act upon; 3) if the service units *were* activated while they were already running - remember, there is `RemainAfterExit= yes` here - then it would be an error for `systemd` to re-activate the service units anyway.  Presumably, `systemd` would not normally do that - and we are searching for, what is presumably, an elusive bug here.

Please keep in mind, that this has been ongoing process, tweaking how `systemd` handles a network interface appearing and disappearing, involving terminology, documentation, and unit file configuration.  Lennart and I have disagreed about the use of the term "dead" - which connotes a "failed/crashed" state - instead of, for instance, "dormant", to reference an inactive but runable state.  And then, I wonder that some developer has not confused "failed" and "inactive" unit states at some point, because of cultural language misinterpretations.  There has also been an issue with the required use of `ConditionPathExists= /sys/class/net/%i` in combination with `BindsTo= sys-subsystem-net-devices-%i.device`, since why should `systemd` have to additionally wait for the creation of a directory by the kernel before is "knows" whether or not a network interface exists?  Isn't that precisely incidental to the systemd device unit itself, "sys-subsystem-net-devices-%i.device"?

And then, where does `systemd` actually check to see if a unit is in "active" state, before deciding to "replace/restart" that same unit?  That just seems like a bug, on its face.  Hopefully, someone more familiar with the `systemd` state engine can explain what is going-on there, and why or how this "Trying to enqueue job sit1@enp6s0.service/restart/replace" state was triggered, and what it represents.

**Comment by yuwata** [comment]
> There has also been an issue with the required use of `ConditionPathExists= /sys/class/net/%i`

I do not think it is necessary, when it has `Wants=` and `After=` for the device unit.

Anyway, please provide details (whole relevant service units, debugging logs, output of `systemctl show` for the relevant services), and minimal reproducer.
Otherwise, there is mostly nothing we can do.

**Comment by thx1111** [comment]
> I do not think it is necessary, when it has Wants= and After= for the device unit.

Things break without it.

> Otherwise, there is mostly nothing we can do.

I understand - but - first, to avoid wading through a lot of noisy log files and replicating my entire system setup, I am asking that someone please explain the trigger source and meaning of the "Trying to enqueue job [sit1@enp6s0.service](mailto:sit1@enp6s0.service)/restart/replace" state.  We don't want to be "looking for a needle in a haystack."

The whole point of running `systemd` with debug reporting was to identify the source of the "sigterm".  And yet, despite having finally found a way to run the network configuration units  with "debug", we still have not identified the original source of this eventual "kill" signal.

There must be someone who is familiar with the `systemd` state engine?

Is this an actual "Restart"?  What is the "Replace" state?

`man systemd.unt` tells me:
```
OnSuccessJobMode=, OnFailureJobMode=
           Takes a value of "fail", "replace", "replace-irreversibly", "isolate", "flush", "ignore-dependencies" or "ignore-requirements". Defaults to "replace". Specifies how the units listed in OnSuccess=/OnFailure= will be
           enqueued. See systemctl(1)'s --job-mode= option for details on the possible values. If this is set to "isolate", only a single unit may be listed in OnSuccess=/OnFailure=.

           Added in version 209.
```
`sit1@enp6s0.service` does not make any reference to `OnSuccess=` or `OnFailure=`, so I do not understand how the "Replace" state might be invoked here.

`man systemctl` tells me:
```
--job-mode=
           When queuing a new job, this option controls how to deal with already queued jobs. It takes one of "fail", "replace", "replace-irreversibly", "isolate", "ignore-dependencies", "ignore-requirements", "flush",
           "triggering", or "restart-dependencies". Defaults to "replace", except when the isolate command is used which implies the "isolate" job mode.
...
           If "replace" (the default) is specified, any conflicting pending job will be replaced, as necessary.
...
```
That seems to suggest that "something" enqueued `sit1@enp6s0.service`, even though it was already running.  What - precisely - was that "something"?  It appears that the service unit is being enqueued *twice* when the network interface is created by the kernel.  Why *twice*?

This could be the exact same mechanism which is also causing `dhcp6c-enp4s0@enp6s0.service` to be "mysteriously" Stopped and Restarted.  The log also shows, similarly, on an already running service unit:
```
 systemd[1]: dhcp6c-enp4s0@enp6s0.service: Trying to enqueue job dhcp6c-enp4s0@enp6s0.service/restart/replace
 systemd[1]: dhcp6c-enp4s0@enp6s0.service: Installed new job dhcp6c-enp4s0@enp6s0.service/restart as 7984
 systemd[1]: dhcp6c-enp4s0@enp6s0.service: Enqueued job dhcp6c-enp4s0@enp6s0.service/restart as 7984
...
 systemd[1]: dhcp6c-enp4s0@enp6s0.service: Changed running -> stop
```
Again, why?
